File: lib/Queue.rb
class Queue

  def initialize size
    @queue = Array.new size
    @upper_bound = size-1
    @front = 0
    @rear = -1
  end

  def enqueue value
    raise "Queue is full" if @rear >= @upper_bound
    @queue[@rear+=1] = value
  end

  def dequeue
    raise 'Queue is empty' if @front > @rear
    @front+=1
    if @front > @rear
      @front = 0
      @rear = -1
    end
  end

  def front
    @queue[@front]
  end

  def rear
    @queue[@rear]
  end
end

File: public/javascripts/dataStructure/Queue.js
// bounded Queue
function Queue(size) {
  this.queue = new Array(size);
  this.upperBound = size;
  this.frontIndex = 0;
  this.rearIndex  = -1;
}

Queue.prototype.enqueue = function(value) {
  if(this.rearIndex+1 < this.upperBound) {
    this.rearIndex++;
    this.queue[this.rearIndex] = value;
  } else {
    throw new Error('The Queue is full');
  }
};

Queue.prototype.dequeue = function() {
  if(this.frontIndex <= this.rearIndex) {
    this.frontIndex++;
  } else {
    throw new Error('The Queue is empty');
  }
};

Queue.prototype.rear = function() {
  return this.queue[this.rearIndex];
};

Queue.prototype.front = function() {
  return this.queue[this.frontIndex];
};
